aws/resource_aws_eip.go

```diff
@@ -181,16 +181,20 @@ func resourceAwsEipRead(d *schema.ResourceData, meta interface{}) error {
 		}
 	}
 
-	// Verify AWS returned our EIP
-	if len(describeAddresses.Addresses) != 1 ||
-		domain == "vpc" && *describeAddresses.Addresses[0].AllocationId != id ||
-		*describeAddresses.Addresses[0].PublicIp != id {
-		if err != nil {
-			return fmt.Errorf("Unable to find EIP: %#v", describeAddresses.Addresses)
+	var address *ec2.Address
+
+	// Sometimes AWS returns EIPs that *aren't* the EIP we're looking for
+	// so we loop over the returned addresses to see if it's in the list of results
+	for _, addr := range describeAddresses.Addresses {
+		if (domain == "vpc" && *addr.AllocationId == id) || *address.PublicIp == id {
+			address = addr
+			break
 		}
 	}
 
-	address := describeAddresses.Addresses[0]
+	if address == nil {
+		return fmt.Errorf("Unable to find EIP: %#v", describeAddresses.Addresses)
```

Comment:
Instead of returning an error in these cases, we generally prefer to remove the resource from the Terraform state to trigger a resource recreation:

```go
if address == nil {
	log.Printf("[WARN] EIP %q not found, removing from state", d.Id())
	d.SetId("")
	return nil
}
```